### ISSUE TITLE
how to compute gradient in ParserState.java

### ISSUE THREAD

**Issue by hitxujian** [percyliang/sempre#146]
all what i asked in class **ParserState.java,**
1、**what's the meaning of member variable  "**customExpectedCounts**"? custom distribution is what？**
(gloss = "Use a custom distribution for computing expected counts")
public CustomExpectedCount customExpectedCounts =CustomExpectedCount.NONE;
2、 about function **computeExpectedCounts(),**the output para **_counts_** is gradient in current example?
/**
   * Fill |counts| with the gradient with respect to the derivations
   * according to a standard exponential family model over a finite set of derivations.
   * Assume that everything has been executed, and compatibility has been computed.
   */
public static void computeExpectedCounts(List<Derivation> derivations, Map<String, Double> counts)
 
**_"standard exponential family " means what? can you give some explanations or papers about it?_**

3、funciton **_computeExpectedCounts()_**,i pass params _**CustomExpectedCount.Top**_, then i know how to fill array _trueScores_ and _predScores_,but **_incr_** variable is do what?
"
double incr = trueScores[i] - predScores[i];
if (incr == 0)
		continue;
deriv.incrementAllFeatureVector(incr, counts);
"

**Comment by ppasupat** [comment]
SEMPRE updates the parameters using gradient ascent like this:
* If a derivation gets the correct answer, increase the parameter weights of all its features.
* If a derivation gets an incorrect answer, decrease the parameter weights of all its features.

The net amount of increments and decrements for each feature is called "expected count" in the code.

By default the expected counts are computed from the gradient as described in [the paper](http://www.cs.stanford.edu/people/ppasupat/resource/ACL2015-paper.pdf). The probability distribution is sometimes called softmax regression which belongs to the exponential family; you can read more about it [here](http://cs229.stanford.edu/notes/cs229-notes1.pdf).

1. Instead of following the gradient, there are other schemes for setting the expected counts. This is controlled by the `customExpectedCounts` option. For example, `customExpectedCounts = TOP` only increase feature weights of the *first* (instead of all) correct derivation and decrease feature weights of the *first* incorrect derivation.

2. As explained above, `counts` will be filled by the gradient by default, but `customExpectedCounts` could change that behavior.

3. `incr` is the net count (increment minus decrement) as explained above

**Comment by hitxujian** [comment]
thanks ppasupat for your detailed and fast explanation! i have two problems,first is my understanding after reading your answers and supplement papers： 
1、
>>
one Example(or one sentence),there are multiply derivations(one derivation maybe right or wrong) .
then mark  **_the net amount of increments and decrements of features_**  as **_response variable_** ,
and mark **_the  input Example_** as **_input variable_** .
>>
after that , make some _**relation**_ between the values of _**the response variable**_ and _**the input variable**_ .
>>
because _**the response variable**_ is a random variable,so they meet some _**distribution**_,and the distribution can be specified by param _**customExpectedCounts**_ .
>>
so this is very similar to logistic regression, which _**fix**_ x , y meets Binomial distribution
2、
second is in your answer "By default _**the expected counts are computed from the gradient**_ as described in the paper",but that paper does not say that,or maybe i can't find?

**Comment by hitxujian** [comment]
is this mean 
>>given Example(natural language) , the **_net amount of increments and decrements for each feature_** meets some distribution . 
which is very similar to logistic regression: given x , y meets Binomial distribution, because y is a distribution,so there are some params,and the params of y has some relation with fixed x,and what we need is build this relation!
is this right??thanks

**Comment by ppasupat** [comment]
The increments and decrements do not follow a distribution.

Given an example x, SEMPRE generates candidate Derivations z, and then define a **distribution** on the Derivations as

![codecogseqn](https://cloud.githubusercontent.com/assets/5135317/26288506/af301122-3e48-11e7-9ede-307ae3e85d7f.png)

where ϕ(z,x) is the feature vector and θ is the parameter vector.

If the correct denotation is y, we want to maximize the probability of getting y. This is the same as summing the probabilities of all Derivations z that execute to y:

![codecogseqn2](https://cloud.githubusercontent.com/assets/5135317/26288532/4d16e334-3e49-11e7-9f64-81386468f5d7.png)

We want to find θ that makes the quantity above large. To apply gradient ascent, we take the gradient with respect to θ. That's how we get the increments and decrements for each feature.

We found that `customExpectedCounts = TOP` is better than the values from the gradient in practice. I currently don't have a solid justification.
